cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java

```diff
@@ -227,6 +228,64 @@ private void isolate(String node, Set<String> addedAlready) {
   }
 
   /**
+   * This method first split the Dag based on the input set of conditions.
+   * The individual splitted dag then further split by using {@link #split} method based on the reduce nodes.
+   * @param conditionNodes Set of condition nodes based on which dag is to be pre-split.
+   * @return the subdags.
+   */
+  public List<Dag> splitByConditions(Set<String> conditionNodes) {
+    List<Dag> dags = new ArrayList<>();
+    Set<String> parentStopperNodes = Sets.union(sources, conditionNodes);
+    Set<String> childStopperNodes = Sets.union(sinks, conditionNodes);
+    for (String condition : conditionNodes) {
+      Set<String> parentNodes = parentsOf(condition, parentStopperNodes);
+      // If the parentNodes only contain conditions, then do not add create subdag for them.
+      // This particular subdag will be created when we go from parent condition to child in the for loop below.
+      if (!Sets.difference(parentNodes, conditionNodes).isEmpty()) {
+        dags.add(createSubDag(parentNodes));
+      }
+      // For child we need to add two sub-dags corresponding to the true and false branch
+      // Condition node has at least one and at max two output nodes
+      Set<String> outputs = getNodeOutputs(condition);
+      for (String output : outputs) {
```

Comment:
If we look at condition parents above, then look at children here, doesn't that mean we can add duplicate subdags? For example:

```
c1 -> source -> c2 -> transform -> c3 -> sink
``` 
Won't it create:
```
c1 -> source -> c2 (when condition == c1)
c1 -> source -> c2 (when condition == c2)
c2 -> transform -> c3 (when condition == c2)
c2 -> transform -> c3 (when condition == c3)
c3 -> sink (when condition == c3)
```
Maybe you could only look at parents and then handle the remaining nodes after the loop?